### PR TITLE
feat(lambda-layer): add pipeline to build Lambda layer in v3

### LINKED ISSUE #3859: Maintenance: Create a Powertools Lambda layer for each Python version
### Why is this needed?

Powertools is growing very fast and we are adding more features day by day. Some of these features require compiled libraries and those libraries must be compiled for the specific version of Python. An example of this is aws-encryption-sdk which brings cryptography as a dependency. We had an issue (https://github.com/aws-powertools/powertools-lambda-python/issues/3852) recently and we need to implement it to avoid more problems in the future.

### Which area does this relate to?

Other

### Solution

Build the Powertools Lambda layer for the specific version of Python.

### Acknowledgment

- [X] This request meets [Powertools for AWS Lambda (Python) Tenets](https://docs.powertools.aws.dev/lambda/python/latest/#tenets)
- [ ] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Java](https://github.com/aws-powertools/powertools-lambda-java/), [TypeScript](https://github.com/aws-powertools/powertools-lambda-typescript/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### PR DESCRIPTION
**Pull request by leandrodamascena**
**Issue number:** #3859 

### Changes

This PR creates the v3 pipeline to accommodate the build and deployment of the Powertools Lambda layer in Python versions from 3.8 to 3.12, and also:

- Adds a section to explain how to build and deploy the layers
- Modifies the files to use the new layer format
- Updates the homepage to display all the layer versions according to the Python versions.

### New names

The names of the Lambda Layer and SAR will be:

**Powertools V3 Layer name:** arn:aws:lambda:{region}:017000801446:layer:AWSLambdaPowertoolsPythonV3-python{pythonversion}-{x86/arm64}:layer-version

**Powertools V3 SAR name:** aws-lambda-powertools-python-layer-v3-python{version}-{x86/arm64}

### User experience

image

image

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by leandrodamascena** [comment]
Missing items:

- [ ] SAR v3
- [ ] Modify CDK Layer Stack
- [ ] Tests with beta accounts
- [ ] Tests with documentation ARN replace

File: docs/automation.md
Comment (leandrodamascena) [inline docs/automation.md]: I would like your review here @heitorlessa and @dreamorosi! I left a few lines commented in the text to see the before and after.

File: docs/index.md
Comment (leandrodamascena) [inline docs/index.md:170]: I would like your review here @heitorlessa and @dreamorosi! Please run the documentation locally and see.

**Review by leandrodamascena — COMMENTED** [review]
I need an initial review @dreamorosi and @heitorlessa! Please do not merge this PR.

**Comment by leandrodamascena** [comment]
The pipeline is working as expected on another repository. Once we merge this PR then the next step is to run a dry-run/beta deployment with V3 on our beta account.
image

File: docs/automation.md
Comment (heitorlessa) [reply docs/automation.md]: Sure!!

Comment (heitorlessa) [inline docs/automation.md]:
```suggestion
## Lambda layer pipeline
```

suggestion: make the title easier to grasp and deep link

File: docs/automation.md
Comment (heitorlessa) [inline docs/automation.md:91]: important: missing prelude; explain what Lambda Layer is and why we need a pipeline.

File: docs/automation.md
Comment (heitorlessa) [inline docs/automation.md]: suggestion: break this up as you're assuming whoever reading knows the intricacies of all that.

> Questions in my head for the non-initiated (new maintainers)

* Why do we build a Layer for each Python version?
* Why do use QEMU for emulating ARM64?
* What's a CDK Asset? can we use a different name?

You could also use sub-heading to help split context and how it's done

File: docs/automation.md
Comment (heitorlessa) [inline docs/automation.md]:
```suggestion
```

question: leftover?

File: docs/automation.md
Comment (heitorlessa) [inline docs/automation.md]:
```suggestion
```

question: leftover?

File: docs/automation.md
Comment (heitorlessa) [inline docs/automation.md]:
```suggestion
Next, we deploy these CDK Assets to the beta account across all AWS regions. Once the beta deployment is complete, we run:
```

suggestion: `execute` word can have negative connotations.

File: docs/automation.md
Comment (heitorlessa) [inline docs/automation.md]:
```suggestion
* **Successful?**: Deploy previous CDK Asset to production across all regions
* **Failure?**: Halt pipeline to investigate and remediate issues before redeploying
```

suggestion: only dropping repetitive words

File: docs/automation.md
Comment (heitorlessa) [inline docs/automation.md]: suggestion: merge with the previous paragraph.

Question to ask in line editing -- if you drop this sentence, does it remove any context?

**Review by heitorlessa — CHANGES_REQUESTED** [review]
Looking SUUUUUUPER good! I'll continue after lunch - dropping suggestions and questions for the docs part.

File: docs/automation.md
Comment (leandrodamascena) [reply docs/automation.md]: Just to compare the before and after.. Removing that.

File: docs/automation.md
Comment (leandrodamascena) [reply docs/automation.md]: Yep! Make sense!

File: docs/automation.md
Comment (leandrodamascena) [reply docs/automation.md]: Just to compare the before and the after. Removing.

File: docs/automation.md
Comment (heitorlessa) [inline docs/automation.md]: suggestion: you can use a feature named `subgraph` to group those activities to make it easier for cognitive load. As a byproduct, it also makes it easier for the next maintainer looking to edit/extend (added space for cognitive load too). 

**After**

image

**Before**

image


```suggestion
```mermaid
graph LR
    Fetch[Fetch PyPi release] --> P38[<strong>Python 3.8</strong>]
    Fetch --> P39[<strong>Python 3.9</strong>]
    Fetch --> P310[<strong>Python 3.10</strong>]
    Fetch --> P311[<strong>Python 3.11</strong>]
    Fetch --> P312[<strong>Python 3.12</strong>]
    
    subgraph build ["LAYER BUILD"]
      P38 --> P38x86[build x86_64]
      P38 --> P38arm64[build arm64]
      
      P39 --> P39x86[build x86_64]
      P39 --> P39arm64[build arm64]

      P310 --> P310x86[build x86_64]
      P310 --> P310arm64[build arm64]

      P311 --> P311x86[build x86_64]
      P311 --> P311arm64[build arm64]

      P312 --> P312x86[build x86_64]
      P312 --> P312arm64[build arm64]

      P38x86 --> CDKP1[CDK Asset]
      P38arm64 --> CDKP1[CDK Asset]

      P39x86 --> CDKP2[CDK Asset]
      P39arm64 --> CDKP2[CDK Asset]

      P310x86 --> CDKP3[CDK Asset]
      P310arm64 --> CDKP3[CDK Asset]

      P311x86 --> CDKP4[CDK Asset]
      P311arm64 --> CDKP4[CDK Asset]

      P312x86 --> CDKP5[CDK Asset]
      P312arm64 --> CDKP5[CDK Asset]
    end
    
    subgraph beta ["BETA (all regions)"]
      CDKP1 --> DeployBeta[Deploy to Beta]
      CDKP2 --> DeployBeta
      CDKP3 --> DeployBeta
      CDKP4 --> DeployBeta
      CDKP5 --> DeployBeta

      DeployBeta --> RunBetaCanary["Beta canary tests<br> <i>(all assets)</i>"]
    end

    subgraph prod ["PROD (all regions)"]
        RunBetaCanary---|<strong>If successful</strong>|DeployProd[Deploy to Prod]
        DeployProd --> RunProdCanary["Prod canary tests<br> <i>(all assets)</i>"]
    end
```
```

File: docs/automation.md
Comment (heitorlessa) [inline docs/automation.md]:
```suggestion
```

leftover, we can get rid of it now :)

**Review by heitorlessa — CHANGES_REQUESTED** [review]
fantastic work with the graph - it makes it easy to learn that we build one per each python version + arch now.

I've made a suggestion to use [`subgraph` feature in mermaid](https://mermaid.js.org/syntax/flowchart.html#subgraphs) so it can reduce cognitive load by breaking into 3 large groups.

**Before**

image


**After**

image

File: layer_v3/layer/canary/app.py
Comment (heitorlessa) [inline layer_v3/layer/canary/app.py:12]: remark: subtle but incredibly important attention to detail to test non-Python dependencies <3

Thank you!!!

File: layer_v3/pyproject.toml
Comment (heitorlessa) [inline layer_v3/pyproject.toml]:
```suggestion
authors = ["Powertools for AWS Maintainers <aws-powertools-maintainers@amazon.com>"]
```

Might be worth checking for other entries with DevAx instead of Powertools maintainers team

File: layer_v3/sar/template.txt
Comment (heitorlessa) [inline layer_v3/sar/template.txt:8]: note: historically, SAR didn't have MIT-0 (MIT with no attribution) so @am29d left it as Apache 2.0 back in the days.

File: pyproject.toml
Comment (heitorlessa) [inline pyproject.toml:3]: note: FREAKING YEAH!!!

File: layer_v3/README.md
Comment (heitorlessa) [inline layer_v3/README.md]: question: is this correct?

File: layer_v3/layer/canary/app.py
Comment (leandrodamascena) [reply layer_v3/layer/canary/app.py:12]: I wanted to make sure that utilities that rely on extra dependencies won't fail!

Attention to detail? Yes, I learned from you how to always raise the bar ❤️

File: examples/metrics_datadog/sam/template.yaml
Comment (heitorlessa) [inline examples/metrics_datadog/sam/template.yaml:28]: remark: nice attention to detail!!!

File: examples/homepage/install/x86_64/amplify.txt
Comment (heitorlessa) [inline examples/homepage/install/x86_64/amplify.txt]:
```suggestion
? Enter up to 5 existing Lambda layer ARNs (comma-separated): arn:aws:lambda:eu-central-1:017000801446:layer:AWSLambdaPowertoolsPythonV3-312:1
```

question: version 1 like others?

File: examples/homepage/install/x86_64/amplify.txt
Comment (heitorlessa) [inline examples/homepage/install/x86_64/amplify.txt]:
```suggestion
? Enter up to 5 existing Lambda layer ARNs (comma-separated): arn:aws:lambda:eu-central-1:017000801446:layer:AWSLambdaPowertoolsPythonV3-312:1
```

question: version 1 like others?

File: examples/homepage/install/arm64/terraform.tf
Comment (heitorlessa) [inline examples/homepage/install/arm64/terraform.tf]:
```suggestion
  layers        = ["arn:aws:lambda:{region}:017000801446:layer:AWSLambdaPowertoolsPythonV3-Arm64-312:1"]
```

question: version 1 like others? I suspect it'll be fixed as part of the release like always but just in case :D

File: examples/homepage/install/arm64/pulumi_arm64.py
Comment (heitorlessa) [inline examples/homepage/install/arm64/pulumi_arm64.py]:
```suggestion
            ":017000801446:layer:AWSLambdaPowertoolsPythonV3-Arm64-312:1",
```

question: version 1 like others?

File: examples/homepage/install/arm64/amplify.txt
Comment (heitorlessa) [inline examples/homepage/install/arm64/amplify.txt]:
```suggestion
? Enter up to 5 existing Lambda layer ARNs (comma-separated): arn:aws:lambda:eu-central-1:017000801446:layer:AWSLambdaPowertoolsPythonV3-Arm64-312:1
```

question: version 1 like others?

File: examples/homepage/install/arm64/amplify.txt
Comment (heitorlessa) [inline examples/homepage/install/arm64/amplify.txt]:
```suggestion
? Enter up to 5 existing Lambda layer ARNs (comma-separated): arn:aws:lambda:eu-central-1:017000801446:layer:AWSLambdaPowertoolsPythonV3-Arm64-312:1
```

question: version 1 like others?

File: docs/index.md
Comment (heitorlessa) [reply docs/index.md:170]: @leandrodamascena to check rendering or something else? 

I just made comment on some leftovers

File: .github/workflows/reusable_deploy_v3_layer_stack.yml
Comment (heitorlessa) [inline .github/workflows/reusable_deploy_v3_layer_stack.yml:3]: question: should we update it with more info on arch etc?

File: .github/workflows/publish_v3_layer.yml
Comment (heitorlessa) [inline .github/workflows/publish_v3_layer.yml:3]: question: is this still valid?

**Review by heitorlessa — CHANGES_REQUESTED** [review]
reviewed every other file and left a few questions to what I think are leftovers

File: layer_v3/README.md
Comment (leandrodamascena) [reply layer_v3/README.md]: Missing the new argument: `pythonVersion`. Added!

File: .github/workflows/publish_v3_layer.yml
Comment (leandrodamascena) [reply .github/workflows/publish_v3_layer.yml:3]: Yes, this is. Do you have any specific question? Because this is valid.

File: .github/workflows/reusable_deploy_v3_layer_stack.yml
Comment (leandrodamascena) [reply .github/workflows/reusable_deploy_v3_layer_stack.yml:3]: Good catch, added!

File: layer_v3/sar/template.txt
Comment (leandrodamascena) [reply layer_v3/sar/template.txt:8]: Thanks for bring more context.

**Review by leandrodamascena — COMMENTED** [review]
Thanks for the review @heitorlessa! I'm working on the last 2 items before ask for another review.

File: examples/homepage/install/arm64/amplify.txt
Comment (leandrodamascena) [reply examples/homepage/install/arm64/amplify.txt]: I had to update this with the new name format we decided on. It's up to date.

File: examples/homepage/install/arm64/amplify.txt
Comment (leandrodamascena) [reply examples/homepage/install/arm64/amplify.txt]: I had to update this with the new name format we decided on. It's up to date.

File: examples/homepage/install/x86_64/amplify.txt
Comment (leandrodamascena) [reply examples/homepage/install/x86_64/amplify.txt]: I had to update this with the new name format we decided on. It's up to date.

File: examples/homepage/install/arm64/pulumi_arm64.py
Comment (leandrodamascena) [reply examples/homepage/install/arm64/pulumi_arm64.py]: I had to update this with the new name format we decided on. It's up to date.

File: examples/homepage/install/arm64/terraform.tf
Comment (leandrodamascena) [reply examples/homepage/install/arm64/terraform.tf]: I had to update this with the new name format we decided on. It's up to date.

File: layer_v3/pyproject.toml
Comment (leandrodamascena) [reply layer_v3/pyproject.toml]: Good catch! Fixed

File: docs/automation.md
Comment (leandrodamascena) [reply docs/automation.md]: This is completely amazing @heitorlessa! Thanks for this suggestion!

File: docs/automation.md
Comment (leandrodamascena) [reply docs/automation.md]: Good catch @heitorlessa! I think using sub-heading make this too extensive, but I added some bullet points. I also removed QMU because this doesn't bring any value in this section.

Can you pls review?

File: docs/automation.md
Comment (leandrodamascena) [reply docs/automation.md:91]: Added

**Review by heitorlessa — APPROVED** [review]

**Pull request merged**
